src/modules/headset/headset_openxr.c

```diff
@@ -1773,6 +1823,63 @@ static bool openxr_animate(Device device, Model* model) {
   return true;
 }
 
+static bool openxr_animateMSFT(XrControllerModelKeyMSFT modelKey, Model* model) {
+  XrControllerModelNodePropertiesMSFT nodeProperties[16];
+  XrControllerModelPropertiesMSFT properties = {
+    .type = XR_TYPE_CONTROLLER_MODEL_PROPERTIES_MSFT,
+    .nodeCapacityInput = COUNTOF(nodeProperties),
+    .nodeProperties = nodeProperties,
+  };
+
+  if (XR_FAILED(xrGetControllerModelPropertiesMSFT(state.session, modelKey, &properties))) {
+    return false;
+  }
+
+  XrControllerModelNodeStateMSFT nodeStates[16];
+  XrControllerModelStateMSFT modelState = {
+    .type = XR_TYPE_CONTROLLER_MODEL_STATE_MSFT,
+    .nodeCapacityInput = COUNTOF(nodeStates),
+    .nodeStates = nodeStates,
+  };
+
+  if (XR_FAILED(xrGetControllerModelStateMSFT(state.session, modelKey, &modelState))) {
+    return false;
+  }
+
+  lovrModelResetNodeTransforms(model);
+
+  for (uint32_t i = 0; i < modelState.nodeCountOutput; i++) {
+    const char* name = nodeProperties[i].nodeName;
+    ModelData* data = lovrModelGetInfo(model)->data;
+    uint64_t nodeIndex = map_get(&data->nodeMap, hash64(name, strlen(name)));
+    lovrCheck(nodeIndex != MAP_NIL, "ModelData has no node named '%s'", name);
+
+    float position[4], scale[4], rotation[4];
+    lovrModelGetNodeTransform(model, nodeIndex, position, scale, rotation, ORIGIN_ROOT);
+
+    vec3_add(position, (vec3)&nodeStates[i].nodePose.position);
+    quat_mul(rotation, rotation, (quat)&nodeStates[i].nodePose.orientation);
```

Comment:
Based on the wording of the extension

> [the pose is] the XrPosef of the node in it’s parent node space

I would have expected the pose in the nodeState to overwrite the node's local pose, not get added as an offset.  i.e. `lovrModelSetNodeTransform(model, nodeIndex, &nodeStates[i].nodePose.position.x, ...)`).

Reply:
yes, that's true. I think I got a bit confused when I first looked at the way the transforms are stored and thought these were absolute, so this was supposed to apply them locally bet that was all wrong.

This is where the float[3]/float[4] thing comes in, passing the position straight in gives this warning:

```
/home/s-ol/Documents/other/lovr/src/modules/headset/headset_openxr.c: In function ‘openxr_animateMSFT’:
/home/s-ol/Documents/other/lovr/src/modules/headset/headset_openxr.c:1867:5: warning: ‘lovrModelSetNodeTransform’ accessing 16 bytes in a region of size 4 [-Wstringop-overflow=]
 1867 |     lovrModelSetNodeTransform(model, nodeIndex, &nodeStates[i].nodePose.position.x, NULL, rotation, 1);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/s-ol/Documents/other/lovr/src/modules/headset/headset_openxr.c:1867:5: note: referencing argument 3 of type ‘float[4]’
/home/s-ol/Documents/other/lovr/src/modules/headset/headset_openxr.c:1867:5: note: referencing argument 4 of type ‘float[4]’
/home/s-ol/Documents/other/lovr/src/modules/headset/headset_openxr.c:1867:5: note: referencing argument 5 of type ‘float[4]’
```

It's not pretty, but making a copy with `vec3_init` works.